Add DerivativeFactory contract as `Creator` in DVM (#115)
UMAprotocol/UMIPs: UMIPs/umip-32.md

# Headers

| UMIP-32  |                                                                                                                                          |
|------------|------------------------------------------------------------------------------------------------------------------------------------------|
| UMIP Title | Register new financial contract with DVM                                                                                                |
| Authors    | Iliyan Iliev (iliyan.iliev@jarvis.exchange) |
| Status     | Draft                                                                                                                                    |
| Created    | December 29, 2020      

# Summary

> Due to necessity and after discussing with the UMA team their protocol was forked. This fork allowed for addition of a few functionalities which were important building blocks of the project. Some of the unnecessary functionalities were removed and parts of the codebase were split in a helper library which allows gas optimization when deployment of the contracts without surpassing the gas limit. The newly updated `DerivativeFactory` contract inherits most of the original functionalities, but also adds specific ones for the needs of the project.

Author: Done in the refactored version

Once registered with the DVM, our `DerivativeFactory` contract will use the liquidation/dispute system set up by UMA, while maintaining the flexibility of deploying new derivative contracts (creating new synthetic assets) or deploying a derivative and linking it an already existing synthetic asset.

# Motivation

The forking of the UMA contracts was a neccessary step in order to align the `DerivativeFactory.sol` with the needs of our project. It also allowed us to go faster with the development as otherwise we would have had to add tons of code and complexity to our code base.

> Making some modifications to the contract allowed us to achieve lower gas consumption for deployment, add the functionality to link newly deployed derivatives to an already existing synthetic tokens and to also deploy new derivatives thus creating new synthetic tokens.

Author: Information has been included with the latest commit

- ``PerpaetualPoolParty.sol` is a derived contract that inherit from `PerpetualLiquidatablePoolParty.sol`, `PerpetualPositionManagerPoolParty.sol` and  `FeePayerPoolParty.sol`. Each one of These has its own library for gas optimization:
`PerpetualLiquidatablePoolPartyLib.sol`, `PerpetualPositionManagerPoolPartyLib.sol`, `FeePayerPoolPartyLib.sol`
- The derivative contract has a whitelist which prevents anyone but Synthereums pool to mint synthetic assets; this prevent an attacker to "spam" our protocol by increasing the GCR.

# Implementation and code base

The `DerivativeFactory` contracts can be found here are available for a review by anyone [here](https://gitlab.com/jarvis-network/apps/exchange/mono-repo/-/tree/feature/uma-integration-part-2/libs/contracts/contracts).

The `PerpetualPoolParty` contracts can be found and are available for a review by anyone [here](https://gitlab.com/jarvis-network/apps/exchange/UMAprotocol/-/tree/jarvis-dev/for-publish/0.3.x/packages/core/contracts).

# Security considerations

> The forked code has been officially audited by a third party company (Halborn), whit whom we are working on security across our whole project. The security audit can be seen [here](https://gitlab.com/jarvis-network/apps/exchange/mono-repo/-/blob/dev/docs/security-audits/jarvis-perpetualpoolparty-halborn-audit.pdf)

Author: Thank you! Still there is a lot more to go

Two low impact issues were found by the auditors and will be addressed as soon as possible. 

> However the protocol itself (Synthereum) hasn't been audited yet, so there is still the chance of issues arrising.To mitigate all risks, we have opted for a very soft and capped launch (only 20k usd worth of asset will be able to be minted as well as having the `emergencyShutdown` functionality enabled.

Author: Included information about that in the refactored version

# Headers

| UMIP-32  |                                                                                                                                          |
|------------|------------------------------------------------------------------------------------------------------------------------------------------|
| UMIP Title | Add `DerivativeFactory` as `Creator` in DVM registry                                                                                                |
| Authors    | Iliyan Iliev (iliyan.iliev@jarvis.exchange) |
| Status     | Draft                                                                                                                                    |
| Created    | December 29, 2020

# Summary

> Due to necessity and after discussing with the UMA team their protocol was forked and some contracts were modified to fit the needs of our protocol. The contracts modified were `PerpetualCreator.sol`, which is now called `PerpetualPoolPartyCreator.sol` and `Perpetual.sol` contract, which is now called `PerpetualPoolParty.sol`. The `DerivativeFactory.sol` is a derived contract of `PerpetualPoolPartyCreator.sol` that restrict the access for deployment to only our deployer contract. An overview of the modifications done can be seen below:

Author: Addressed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b)

> Restricted access for deposit/withdraw and create/redeem was added in `Perpetual.sol` (`PerpetualPoolParty.sol`) in order to allow access to those operations only to our Synthereum protocol.

Author: Addressed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b)

> Due to the design of our protocol we don't need `fundingRate` from `Perpetual.sol` for our perpetual assets to incentivize market actors to help us maintaining a peg, so this functionality was removed in order to optimize the gas consumption. In Synthereum the token sponsor is a contract and any synthetic asset bought on the open market can be burned in exchange of its exact value in collateral against a price feed, which guarantee constant arbitrage between Synthereum and DEXes, thus maintaining the peg.

Reviewer: What is the source for this "price feed"?

Author: @nicholaspai Actually this part was changed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b) as suggested by @mrice32 

However to answer your question:

Currently we are using an off-chain price feed provided by [TraderMade API](https://marketdata.tradermade.com/documentation) and our backend called Validator (KeeperBot) and the user make an agreement on the price the transaction should happen. This price feed is also used by voters if a liquidation/dispute process is started as noted in [UMIP-29](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-29.md). Also in the same UMIP you can find alternative APIs for price feed provision. Due to the possible difference between said price feed and the price of a synthetic asset on a DEX (as example Uniswap) an arbitrage opportunity is created, which incentivizes users to trade the synthetic stocks both on Jarvis Exchange and other DEXes thus maintaining the peg. 

But of course we plan on becoming fully decentralized, which means removing the Validator (KeeperBot). In order for this to happen we'll be moving to an on-chain price feed using Chainlink in the near future. The Validator was created in order to have a working version at the beginning, but the plan always was for this version to be temporary.

Reviewer: > backend called Validator (KeeperBot) and the user make an agreement on the price

I.e. the User is presented an option to sign & submit a transaction that has some implied price, originally sourced from TraderMade?

> Due to the possible difference between said price feed and the price of a synthetic asset on a DEX (as example Uniswap) an arbitrage opportunity is created, which incentivizes users to trade the synthetic stocks both on Jarvis Exchange and other DEXes thus maintaining the peg.

This is a cool feature, I agree it should trade close to the price that the "KeeperBot" continually presents to the user. The main uncertainty is the reliability of the KeeperBot, and by extension TraderMade. But I think this is OK as long as the risks are communicated to your end users.

> But of course we plan on becoming fully decentralized, which means removing the Validator (KeeperBot). In order for this to happen we'll be moving to an on-chain price feed using Chainlink in the near future. 

Sounds good to me. This will be an awesome first implementation of a financial contract that uses Chainlink optimistically for live-prices, and the UMA DVM for ultimate insurance with human discretion!

Author: @nicholaspai To answer your first question: We call it an agreement, however the way it works is the price the user sees in the frontend is the same price that the KeeperBot is receiving from TraderMade API. Once the user triggers a transaction (since we use metasignatures) the message is send to the keeper bot which signs it, then send back to the user and send to the blockchain for validation. Ofc there might be a slight change of price within this period, but due to Forex prices being quite stable it won't affect the overall transaction.

We completely agree that the KeeperBot is a point of failure, that is why our first goal as mentioned would be to move to an on-chain price feed. The KeeperBot was created as an initial mechanism with which to go on Mainnet, but not final ofc.

> Due to the design of our protocol we don't need `fundingRate` from `Perpetual.sol` for our perpetual assets to incentivize market actors to help us maintaining a peg, so this functionality was removed in order to optimize the gas consumption. In Synthereum the token sponsor is a contract and any synthetic asset bought on the open market can be burned in exchange of its exact value in collateral against a price feed, which guarantee constant arbitrage between Synthereum and DEXes, thus maintaining the peg.

Reviewer 2: Some minor edits here, but also a question: can you also mint/create the token at the market price? If not, what's to keep the market price from going _above_ the underlying price feed's price?

```suggestion
All funding rate logic was removed from `PerpetualPoolParty.sol`. In Synthereum, the token sponsor is a contract and any synthetic asset bought/sold on the open market can be created/burned in exchange for its exact value in collateral based on a price feed, which guarantees a perfect arbitrage between Synthereum and DEXs, thus maintaining the peg.
```

Author: Addressed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b)

Author: @mrice32 Regarding your question:

The price feed (TraderMade) which we are using provide us with real-time price feed data, which is constantly updated in our UI. Once the user triggers a transaction from the UI, before it is broadcasted to the blockchain, this transaction passes through our Validator (KeeperBot) which verifies that the price is correct and an agreement is reached between the validator and the user for the tokens to be minted as example at said price. So it is kinda impossible for the market price to go above the underlying price feed's price.

However one note here is that while the transaction is broadcasted to the blockchain and is confirmed the price of an asset can move, which is something that at this point can not be addressed. So basically if the user agrees with the Validator to mint X amount of tokens at price N, by the time the transaction is cleared the price of the asset might be M, meaning that either the user has already won some money on his investment or lost some money depending on where the price went.

Reviewer 2: Makes sense, so to summarize, users can both burn _and create_ tokens at the market price. SGTM.

Author: @mrice32 Yes that is correct :)

> Since our assets are perpetual if we wanted to upgrade or change something in the derivative contract (as example change of collateral) it will require to deploy a new token with the newly deployed derivative, which will create frictions for users, liquidity providers and protocols where our synthetic tokens are integrated, thus we wanted to make derivatives contracts upgradable (by deploying a new one), but still keeping it linked to the already existing perpetual synthetic asset. Those functionalities were changed in the `PerpetualCreator.sol` (`PerpetualPoolPartyCreator.sol`).

Reviewer 2: Some minor grammatical edits. Just so I'm understanding correctly, this is set up so you can deploy a PerpetualPoolParty contract and migrate an existing synthetic token to the new contract? How does this new contract initialize/migrate the old positions to the new contract to ensure all tokens are backed by collateral?

```suggestion
Since our assets are perpetual, if we wanted to upgrade or change something in the derivative contract (change of collateral, for instance) it would require a new perpetual token to be deployed, which will create friction for end-users, liquidity providers, and protocols where our synthetic tokens are integrated. Thus, we wanted to be able to deploy a new derivative contract, but link it to the preexisting perpetual token. This upgradability was added in `PerpetualPoolPartyCreator.sol`.
```

Author: Addressed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b)

> Another modification that was done was splitting some of the functionalities of `Perpetual.sol` (`PerpetualPoolParty.sol`) into separate libraries mentioned below in the technical specifications in order to optimize the gas consumption.

Reviewer 2: ```suggestion
`PerpetualPoolParty.sol` was split into separate libraries in order to optimize the gas consumption. Note: this is detailed below in the Technical Specification section.
```

Author: Addressed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b)

> The `emergencyShutdown` functionality in `Perpetual.sol` (`PerpetualPoolParty.sol`) has been modified to be called not only by UMA DAO, but also by the Jarvis DAO.

Reviewer 2: ```suggestion
The `emergencyShutdown` functionality in `PerpetualPoolParty.sol` has been modified to not only be callable by the UMA DVM, but also by the Jarvis DAO.
```

Author: Addressed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b)

> Once `DerivativeFactory.sol` receives the `Creator` role it will allow every new `PerpetualPoolParty.sol` that is deployed to be automatically registered within `Registry.sol` and use the liquidation/dispute system set up by UMA.

Reviewer 2: ```suggestion
Once `DerivativeFactory.sol` receives the `Creator` role it will register every `PerpetualPoolParty.sol` that is deployed through it with `Registry.sol` so they can request prices from the DVM.
```

Author: Fixed in this [commit](https://github.com/UMAprotocol/UMIPs/pull/115/commits/f1f823bf9f96dc351a0378bd4eebb5e16d9ba67b)

# Motivation

The forking of the UMA contracts was a necessary step in order to align the `DerivativeFactory.sol` with the needs of our project. It also allowed us to go faster with the development as otherwise we would have had to add tons of code and complexity to our code base.
Making some modifications to the contract allowed us to achieve lower gas consumption for deployment, add the functionality to link newly deployed derivatives to an already existing synthetic tokens and to also deploy new derivatives thus creating new synthetic tokens. Adding this functionality provides our users with a better UX, due to the fact that there won't be multiple versions of the same synthetic asset even if a new derivative contract is deployed for that synthetic asset, thus there won't be the need for users or liquidity providers to transfer their funds in order to back another synthetic asset.
Giving `Creator` role to `DerivativeFactory.sol` would allow our synthetic assets to be backed by the UMA liquidation/dispute system, since every newly deployed derivative (`PerpetualPoolParty.sol`) is registered with the DVM.

> # Technical Specification

Reviewer 3: Can you add the addresses of the deployed and whitelisted contracts once they are available?

Author: @smb2796 Yes as soon as they are deployed and verified on Mainnet I'll add them in a separate paragraph

2. `Deployer.sol` calls `createPerpetual` function of `DerivativeFactory.sol` and this function will call `createPerpetual` function of the base contract `PerpetualPoolPartyCreator.sol` that will deploy the new `PerpetualPoolParty.sol`.
3. `Deployer.sol` calls `createPool` function of `PoolFactory.sol` and this function will call `createPool` function of the base contract `PoolCreator.sol` that will deploy the new `Pool.sol`.
4. `Deployer.sol` links the newly deployed `Pool.sol` to the newly deployed `PerpetualPoolParty.sol` and assignes the roles in `PerpetualPoolParty.sol` to `Pool.sol`, which are `Admin` and `Pool`.

### Modifications done: 

-  The derivative contract (`PerpetualPoolParty.sol`) has a restricted access for deposit/withdraw and create/redeem, which prevents anyone but Synthereums pool (`Pool.sol`) to mint or redeem synthetic assets; this prevent an attacker to "spam" our protocol by increasing the Global Collateralization Ratio (GCR).
- `DerivativeFactory.sol` contract allow to deploy derivatives (create new synthetic tokens) or deploy a derivative and link it to an already existing synthetic token.

> - `DerivivativeFactory.sol` is the derived contract of `PerpetualPoolPartyCreator.sol`(that uses `PerpetualPoolPartyLib.sol` for gas optimization).  Adding this contract as Creator role in UMA registry, allow the derivatives deployed to integrate the DVM

Reviewer 3: ```suggestion
- `DerivativeFactory.sol` is the derived contract of `PerpetualPoolPartyCreator.sol`(that uses `PerpetualPoolPartyLib.sol` for gas optimization).  Adding this contract as Creator role in UMA registry, allow the derivatives deployed to integrate the DVM.
```

# Headers

> | UMIP-32  |                                                                                                                                          |

Reviewer 3: Please also change the filename to umip-34

```suggestion
| UMIP-34  |                                                                                                                                          |
```
